external/llvm-project/mlir/lib/Dialect/LLVMIR/Transforms/SoftwareBf16.cpp

```diff
@@ -0,0 +1,364 @@
+    Value shifted = rewriter.create<LLVM::ShlOp>(
+        loc, extended, getLlvmI32Const(loc, rewriter, extType, 16));
+    rewriter.replaceOpWithNewOp<LLVM::BitcastOp>(op, destType, shifted);
+
+    return success();
+  }
+};
+
+/// Rewrites truncation to bfloat as a series of integer operations.
+struct SoftwareBF16Trunc : OpRewritePattern<LLVM::FPTruncOp> {
+  explicit SoftwareBF16Trunc(MLIRContext *context)
+      : OpRewritePattern(context) {}
+
+  LogicalResult matchAndRewrite(LLVM::FPTruncOp op,
+                                PatternRewriter &rewriter) const override {
+
+    auto loc = op.getLoc();
+
+    Type srcType = op.getArg().getType();
+    Type destType = op.getRes().getType();
+    Type srcElemType = srcType;
+    if (auto shaped = srcType.dyn_cast<ShapedType>())
+      srcElemType = shaped.getElementType();
+
+    Type f32 = rewriter.getF32Type();
+    if (srcElemType != f32)
+      return failure();
+
+    Type bitcastType = rewriter.getI32Type();
+    if (auto srcShaped = srcType.dyn_cast<ShapedType>())
+      bitcastType = srcShaped.clone(bitcastType);
+
+    Type i16 = rewriter.getIntegerType(16);
+    if (auto destShaped = destType.dyn_cast<ShapedType>()) {
+      if (destShaped.getElementType() != i16)
+        return failure();
+    } else if (destType != i16)
+      return failure();
+
+    // a = bitcast f32 value to i32
+    // b = (a + 32767) << 16
+    // c = ((a << 16) & 1)
+    // d = b + c
+    // truncate (d << 16) to i16 and return this i16
+    Value bitcastop =
+        rewriter.create<LLVM::BitcastOp>(loc, bitcastType, op.getArg());
+    Value constantSixteen = getLlvmI32Const(loc, rewriter, bitcastType, 16);
+    Value shiftValue = rewriter.create<LLVM::LShrOp>(
+        loc, bitcastType, bitcastop, constantSixteen);
+
+    Value constantOne = getLlvmI32Const(loc, rewriter, bitcastType, 1);
+    Value andValue = rewriter.create<LLVM::AndOp>(loc, shiftValue, constantOne);
+
+    Value constantBig = getLlvmI32Const(loc, rewriter, bitcastType, 32767);
+    Value addBigValue =
+        rewriter.create<LLVM::AddOp>(loc, bitcastop, constantBig);
+    Value addValue = rewriter.create<LLVM::AddOp>(loc, andValue, addBigValue);
+
+    Value shiftBeforeTruncValue = rewriter.create<LLVM::LShrOp>(
+        loc, bitcastType, addValue, constantSixteen);
+    Value truncValue =
+        rewriter.create<LLVM::TruncOp>(loc, destType, shiftBeforeTruncValue);
+    rewriter.replaceOp(op.getOperation(), {truncValue});
+
+    return success();
+  }
+};
+
+void replaceBF16InOp(Operation *op, LLVMTypeConverter &converter);
+
+void replaceBF16InBlock(Block &block, LLVMTypeConverter &converter) {
+  for (Operation &op : block.getOperations())
+    replaceBF16InOp(&op, converter);
+}
+
+void replaceBF16InRegion(Region &region, LLVMTypeConverter &converter) {
+  for (Block &block : region.getBlocks())
+    replaceBF16InBlock(block, converter);
+}
+
+void replaceBF16InOp(Operation *op, LLVMTypeConverter &converter) {
```

Comment:
MLIR prefers a worklist over recursion - you can probably merge these into one function by using block and operand worklists